mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td

```diff
@@ -531,7 +523,7 @@ def SYCLConstructorOp : SYCL_Op<"constructor", [CallOpInterface]> {
 
   let arguments = (ins
     Arg<SYCLMemref, "The SYCL object being created", [MemWrite]>:$Dst,
-    Variadic<ConstructorArgs>:$Args,  
+    Arg<Variadic<AnyType>, "The constructor arguments", [MemRead, MemWrite]>:$Args,
```

Comment:
Please remove this change in this PR (this new code has the effect of inhibiting the LICM optimization).

Reply:
We have constructors that take scalar arguments. For example:  
```
    %7 = arith.extui %6 : i32 to i64
    sycl.constructor @id(%8, %7) {MangledFunctionName = @_ZN4sycl3_V12idILi1EEC1ILi1EEENSt9enable_ifIXeqT_Li1EEmE4typeE} : (memref<?x!sycl.id<[1], (!sycl.array<[1], (memref<
1xi64, 4>)>)>, 4>, i64)
```
Here %7 is an integer so it doesn't make sense to state that it has memory effects. Only "memrefs" can read or write the memory pointed to.

I think we might need to create more than one `sycl.constructor` operator, if any SYCL RT constructor takes pointer arguments (other than the `this` pointer). IF that is the case we can then properly attach the memory traits to all pointer arguments.

Reply:
is it possible to have an optional variadic args of pointer types (which we can add the memory effect) follow by another optional variadic args of non-pointer types?

Reply:
I guess we could something similar, but that would impose that no mutable argument could appear after a const one. Not sure we want to do that. WDYT?

Reply:
I think what @whitneywhtsang suggest would be useful. The first arg is the 'this' pointer and we mark that one with [MemWrite]. We know that all of the SYCL RT constructors can only ready the remaining arguments. If we group all of the pointer args in a vararg parameter pack we can them mark them all with [MemRead], the remaining vararg parameter pack contains scalars so no need to mark those arguments with any memory effect trait. 

I am not sure though whether we would be able to have a sycl.constructor with 2 arguments. One of those 2 vararg list might be empty, if we can make the Optional it would work though.

Reply:
Nope, but we could in the future. I'm interested. Do we have any real test in which we're calling the constructor in a loop and this prevents LICM or other optimizations?

Reply:
I'll think of solutions to fix this in a generic fashion. Thanks for pointing out